### ISSUE TITLE
java.io.IOException: Buffer underflow

### ISSUE THREAD

**Issue by enragedginger** [classgraph/classgraph#407]
Hello,

I ran into an issue while using ClassGraph version 4.8.65 with Java 8. ClassGraph is able to find most classes on my classpath, but there is one top-level class in particular for which it always throws a `java.io.IOException: Buffer underflow` error in the verbose output. I cloned down the classgraph repo, and updated the error handling to print the full stacktrace and got this:

```
java.io.IOException: Buffer underflow
at nonapi.io.github.classgraph.fileslice.reader.ClassfileReader.readTo(ClassfileReader.java:209)
at nonapi.io.github.classgraph.fileslice.reader.ClassfileReader.skip(ClassfileReader.java:395)
at io.github.classgraph.Classfile.readMethods(Classfile.java:1463)
at io.github.classgraph.Classfile.<init>(Classfile.java:1663)
at io.github.classgraph.Scanner$ClassfileScannerWorkUnitProcessor.processWorkUnit(Scanner.java:664)
at io.github.classgraph.Scanner$ClassfileScannerWorkUnitProcessor.processWorkUnit(Scanner.java:587)
at nonapi.io.github.classgraph.concurrency.WorkQueue.runWorkLoop(WorkQueue.java:246)
at nonapi.io.github.classgraph.concurrency.WorkQueue.runWorkQueue(WorkQueue.java:161)
at io.github.classgraph.Scanner.processWorkUnits(Scanner.java:342)
at io.github.classgraph.Scanner.performScan(Scanner.java:902)
at io.github.classgraph.Scanner.openClasspathElementsThenScan(Scanner.java:1044)
at io.github.classgraph.Scanner.call(Scanner.java:1079)
at io.github.classgraph.Scanner.call(Scanner.java:78)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
```

I can't post the source code due to IP issues. So I tried to create a test class that resembled my issue as closely as possible, but was unable to reproduce the issue with that class.

I did, however, find 2 publicly available classes which produce the issue:

* `org.apache.datasketches.cpc.CompressionData` from `org.apache.datasketches:datasketches-java-1.1.0-incubating`
* `com.google.thirdparty.publicsuffix.PublicSuffixPatterns` from `com.google.guava:guava:25.0-jre`

**Comment by lukehutch** [comment]
@enragedginger Thanks for the detective work! I'm on it.

**Comment by lukehutch** [comment]
@enragedginger This is fixed, and will be in 4.8.66. The issue was that reading those classes required an unusually large seek, and ClassGraph was only ever doubling the buffer size on buffer underrun, whereas it needed to be able to grow the buffer by an arbitrary power of 2.

I can't push out a release right now because Sonatype is having issues, but you can build from git master if you need an immediate fix. Thanks for reporting the bug!

**Comment by enragedginger** [comment]
Wow. That was fast. Thanks for resolving this so quickly! Let me know when the new build is pushed.

**Comment by lukehutch** [comment]
That's the goal 😀 :

https://github.com/classgraph/classgraph/blob/master/Zero-Bugs-Commitment.md

**Comment by lukehutch** [comment]
Released in 4.8.66. Thanks again!
